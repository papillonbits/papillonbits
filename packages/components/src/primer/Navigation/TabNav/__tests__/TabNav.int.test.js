import {
  ariaAttr,
  defaultItems as items,
  navigationItems,
  mixedEnabledVisibleItems,
  custom,
  regular,
  withButtonActions,
  withTextActions,
  withLinkActions,
  withNavigation,
  withMixedEnabledVisibleItems,
  inactive,
} from './TabNav.int.story'

jest.mock('@papillonbits/library/array', () => ({
  getIndexItems: () => {},
  getIndexItemsWithSelected: () => {},
}))

const libraryArrayMockObject = require('@papillonbits/library/array')

describe('<TabNav />', () => {
  const indexItemsDataObject = [
    { ...items[0], index: 0 },
    { ...items[1], index: 1 },
  ]

  const indexNavigationItemsDataObject = [
    { ...navigationItems[0], index: 0 },
    { ...navigationItems[1], index: 1 },
  ]

  const indexMixedEnabledVisibleItemsDataObject = [
    { ...mixedEnabledVisibleItems[0], index: 0 },
    { ...mixedEnabledVisibleItems[1], index: 1 },
    { ...mixedEnabledVisibleItems[2], index: 2 },
  ]

  const newIndexItemsWithSelectedDataObject = [
    { ...items[0], index: 0, isSelected: true, enabled: true, visible: true },
    { ...items[1], index: 1, isSelected: false, enabled: true, visible: true },
  ]

  const newIndexNavigationItemsWithSelectedDataObject = [
    { ...navigationItems[0], index: 0, isSelected: true, enabled: true, visible: true },
    { ...navigationItems[1], index: 1, isSelected: false, enabled: true, visible: true },
  ]

  const newIndexMixedEnabledVisibleItemsDataObject = [
    { ...mixedEnabledVisibleItems[0], index: 0, isSelected: true, enabled: true, visible: true },
    { ...mixedEnabledVisibleItems[1], index: 1, isSelected: false, enabled: false, visible: true },
    { ...mixedEnabledVisibleItems[2], index: 2, isSelected: false, enabled: false, visible: false },
  ]

  const newItemsDataObject = [
    { ...items[0], isSelected: true, enabled: true, visible: true },
    { ...items[1], isSelected: false, enabled: true, visible: true },
  ]

  const newNavigationItemsDataObject = [
    { ...navigationItems[0], isSelected: true, enabled: true, visible: true },
    { ...navigationItems[1], isSelected: false, enabled: true, visible: true },
  ]

  const newMixedEnabledVisibleItemsDataObject = [
    { ...mixedEnabledVisibleItems[0], isSelected: true, enabled: true, visible: true },
    { ...mixedEnabledVisibleItems[1], isSelected: false, enabled: false, visible: true },
    { ...mixedEnabledVisibleItems[2], isSelected: false, enabled: false, visible: false },
  ]

  const onClickMockFn = jest.fn()

  function mockLibraryForRegularItems() {
    jest.spyOn(libraryArrayMockObject, 'getIndexItems').mockReturnValue(indexItemsDataObject)

    jest.spyOn(libraryArrayMockObject, 'getIndexItemsWithSelected').mockReturnValue(newIndexItemsWithSelectedDataObject)
  }

  function mockLibraryForNavigationItems() {
    jest.spyOn(libraryArrayMockObject, 'getIndexItems').mockReturnValue(indexNavigationItemsDataObject)

    jest.spyOn(libraryArrayMockObject, 'getIndexItemsWithSelected').mockReturnValue(newIndexNavigationItemsWithSelectedDataObject)
  }

  function mockLibraryForMixedEnabledVisibleItems() {
    jest.spyOn(libraryArrayMockObject, 'getIndexItems').mockReturnValue(indexMixedEnabledVisibleItemsDataObject)

    jest.spyOn(libraryArrayMockObject, 'getIndexItemsWithSelected').mockReturnValue(newIndexMixedEnabledVisibleItemsDataObject)
  }

  afterEach(() => jest.clearAllMocks())

  describe('Event', () => {
    test('must return new items without index when regular items and onClick()', () => {
      mockLibraryForRegularItems()
      const mountCustomRender = global.renderMount(custom(items, onClickMockFn)).find('TabNav')
      mountCustomRender.find('nav').first().find('a').first().simulate('click')
      expect(mountCustomRender.props().onClick).toBe(onClickMockFn)
      expect(mountCustomRender.props().onClick).toHaveBeenCalledWith({
        ariaAttr,
        items: newItemsDataObject,
      })
    })

    test('must return new items without index when navigation items and onClick()', () => {
      mockLibraryForNavigationItems()
      const mountCustomRender = global.renderMount(custom(navigationItems, onClickMockFn)).find('TabNav')
      mountCustomRender.find('nav').first().find('a').first().simulate('click')
      expect(mountCustomRender.props().onClick).toBe(onClickMockFn)
      expect(mountCustomRender.props().onClick).toHaveBeenCalledWith({
        ariaAttr,
        items: newNavigationItemsDataObject,
      })
    })

    test('must return new items without index when mixed enabled visuble items and onClick()', () => {
      mockLibraryForMixedEnabledVisibleItems()
      const mountCustomRender = global.renderMount(custom(mixedEnabledVisibleItems, onClickMockFn)).find('TabNav')
      mountCustomRender.find('nav').first().find('a').first().simulate('click')
      expect(mountCustomRender.props().onClick).toBe(onClickMockFn)
      expect(mountCustomRender.props().onClick).toHaveBeenCalledWith({
        ariaAttr,
        items: newMixedEnabledVisibleItemsDataObject,
      })
    })
  })

  describe('Render', () => {
    test('must match regular()', () => {
      mockLibraryForRegularItems()
      expect(global.renderToJSON(regular())).toMatchSnapshot()
    })

    test('must match withButtonActions()', () => {
      mockLibraryForRegularItems()
      expect(global.renderToJSON(withButtonActions())).toMatchSnapshot()
    })

    test('must match withTextActions()', () => {
      mockLibraryForRegularItems()
      expect(global.renderToJSON(withTextActions())).toMatchSnapshot()
    })

    test('must match withLinkActions()', () => {
      mockLibraryForRegularItems()
      expect(global.renderToJSON(withLinkActions())).toMatchSnapshot()
    })

    test('must match withNavigation()', () => {
      mockLibraryForRegularItems()
      expect(global.renderToJSON(withNavigation())).toMatchSnapshot()
    })

    test('must match withMixedEnabledVisibleItems()', () => {
      mockLibraryForRegularItems()
      expect(global.renderToJSON(withMixedEnabledVisibleItems())).toMatchSnapshot()
    })

    test('must match inactive()', () => {
      mockLibraryForRegularItems()
      expect(global.renderToJSON(inactive())).toMatchSnapshot()
    })
  })
})
