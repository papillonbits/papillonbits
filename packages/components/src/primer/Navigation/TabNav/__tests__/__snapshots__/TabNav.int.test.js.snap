// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabNav /> Render must match inactive() 1`] = `
<div
  className="tabnav tabnav-inactive"
>
  <nav
    aria-label="Foo bar"
    className="tabnav-tabs"
  >
    <a
      aria-current="page"
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Foo tab
    </a>
    <a
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Bar tab
    </a>
  </nav>
</div>
`;

exports[`<TabNav /> Render must match regular() 1`] = `
<div
  className="tabnav"
>
  <nav
    aria-label="Foo bar"
    className="tabnav-tabs"
  >
    <a
      aria-current="page"
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Foo tab
    </a>
    <a
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Bar tab
    </a>
  </nav>
</div>
`;

exports[`<TabNav /> Render must match withButtonActions() 1`] = `
<div
  className="tabnav"
>
  <a
    className="btn btn-sm float-right"
    href="#url"
    role="button"
  >
    Button
  </a>
  <nav
    aria-label="Foo bar"
    className="tabnav-tabs"
  >
    <a
      aria-current="page"
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Foo tab
    </a>
    <a
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Bar tab
    </a>
  </nav>
</div>
`;

exports[`<TabNav /> Render must match withLinkActions() 1`] = `
<div
  className="tabnav"
>
  <div
    className="float-right"
  >
    <a
      className="tabnav-extra"
      href="#url"
    >
      Tabnav extra link
    </a>
    <a
      className="tabnav-extra"
      href="#url"
    >
      Tabnav extra link
    </a>
  </div>
  <nav
    aria-label="Foo bar"
    className="tabnav-tabs"
  >
    <a
      aria-current="page"
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Foo tab
    </a>
    <a
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Bar tab
    </a>
  </nav>
</div>
`;

exports[`<TabNav /> Render must match withMixedEnabledVisibleItems() 1`] = `
<div
  className="tabnav"
>
  <nav
    aria-label="Foo bar"
    className="tabnav-tabs"
  >
    <a
      aria-current="page"
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Foo tab
    </a>
    <a
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Bar tab
    </a>
  </nav>
</div>
`;

exports[`<TabNav /> Render must match withNavigation() 1`] = `
<div
  className="tabnav"
>
  <nav
    aria-label="Foo bar"
    className="tabnav-tabs"
  >
    <a
      aria-current="page"
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Foo tab
    </a>
    <a
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Bar tab
    </a>
  </nav>
</div>
`;

exports[`<TabNav /> Render must match withTextActions() 1`] = `
<div
  className="tabnav"
>
  <div
    className="tabnav-extra float-right"
  >
    Tabnav widget text here.
  </div>
  <nav
    aria-label="Foo bar"
    className="tabnav-tabs"
  >
    <a
      aria-current="page"
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Foo tab
    </a>
    <a
      className="tabnav-tab"
      href="#url"
      onClick={[Function]}
    >
      Bar tab
    </a>
  </nav>
</div>
`;
